openedx/core/lib/block_structure/transformer.py

```diff
@@ -125,4 +126,55 @@ def transform(self, usage_info, block_structure):
                 block structure, with already collected data for the
                 transformer, that is to be transformed in place.
         """
-        pass
+        raise NotImplementedError
+
+
+class FilteringTransformerMixin(BlockStructureTransformer):
+    """
+    Transformers may optionally choose to implement this mixin if their
+    transform logic can be broken apart into a lambda for optimization of
+    combined tree traversals.
+
+    For performance reasons, developers should try to implement this mixin
+    whenever possible - with this alternative, traversal of the entire block
+    structure happens only once for all transformers that implement
+    FilteringTransformerMixin.
+    """
+
+    def transform(self, usage_info, block_structure):
+        """
+        This override of the abstract base method allows FilteringTransformers
+        to define transform_block_filter instead of transform.
+        """
+        block_structure.filter_topological_traversal(self.transform_block_filter(usage_info, block_structure))
+
+    @abstractmethod
+    def transform_block_filter(self, usage_info, block_structure):
+        """
+        This is an alternative to the standard transform method.
+
+        Returns a list of filter functions to be used for filtering out
```

Comment:
I ended up going with a list instead of generators for simplicity's sake - I kept running into max_recursion errors when running tests with generators, and it was getting hairy having a generator of generators of lambdas (in the case of user_partition, which calls split_test).